saq/worker.py

```diff
@@ -192,25 +196,37 @@ async def start(self) -> None:
         except asyncio.CancelledError:
             pass
         finally:
-            logger.info("Worker shutting down")
-
-            if self.shutdown:
-                for s in self.shutdown:
-                    await s(self.context)
-
+            logger.info("Working shutting down")
             await self.stop()
 
     async def stop(self) -> None:
         """Stop the worker and cleanup."""
         self.event.set()
-        all_tasks = list(self.tasks)
-        self.tasks.clear()
-        await cancel_tasks(all_tasks)
-
-        if sys.version_info[0:2] < (3, 9):
-            self.pool.shutdown()
-        else:
-            self.pool.shutdown(cancel_futures=True)
+        async with AsyncExitStack() as stack:
+            await stack.enter_async_context(self._stop_lock)
```

Comment:
i don't understand all of this code 205-227, can you explain it to me?

Reply:
- Avoid multiple stop jobs running at once with a lock
- Use a flag to avoid `stop()` cleanup running twice, as `stop()` can trigger another invocation of itself by `start()` by the `Worker.event` shutdown signal.
- Use the new-ish `AsyncExitStack` to cleanly collapse the lock context manager and setting `_stopped` flag on exit into a single `async with` scope.
- Just like before, cancel all tasks both async and threadpool
- But now that we're cancelling tasks (and hence stopping the threadpool) before running shutdown callbacks, we need a new threadpool to run any sync callbacks.

Reply:
Yes of course. To be honest, I can't fault you, I was also taken aback by the exit stack until I got used to reading and using it myself.